### PR TITLE
add node num to RoleMaker; fix recompute Meta Optimizer;  raise error for undefined attr in distributeOptimizer 

### PR DESCRIPTION
**Pull request by mapingshuo**
### PR types

Bug fixes
### PR changes

APIs
### Describe

- add node num to RoleMaker; 
- fix recompute Meta Optimizer; 
- raise error for undefined attr in distributeOptimizer

### CODE REVIEW

File: python/paddle/fleet/base/role_maker.py [hunk 1]
```diff
@@ -286,6 +294,14 @@ def server_num(self):
             self.generate_role()
         return self._trainers_num
 
+    def node_num(self):
+        """
+        return the training node number
+        """
+        if not self._role_is_generated:
+            self.generate_role()
+        return self._node_num
```
Comment (seiriosPlus) [inline python/paddle/fleet/base/role_maker.py:303]: `node_num` define twice?

File: python/paddle/fleet/base/role_maker.py [hunk 2]
```diff
@@ -101,6 +101,14 @@ def server_index(self):
         """
         raise NotImplementedError("Please implement this method in child class")
 
+    def node_num(self):
+        """
+        Get the training node number
+        Returns:
+            int: node num
+        """
+        raise NotImplementedError("Please implement this method in child class")
```
Comment (seiriosPlus) [inline python/paddle/fleet/base/role_maker.py:110]: what is the difference between `node_num` and `worker_num`

Comment (mapingshuo) [reply python/paddle/fleet/base/role_maker.py:110]: > what is the difference between `node_num` and `worker_num`

For the machine with muti-GPU, the `node num` stands for the number of machines, and the `worker_num` stands for the number of GPU cards.

File: python/paddle/fleet/base/role_maker.py (hunk 1 above)
Comment (mapingshuo) [reply python/paddle/fleet/base/role_maker.py:303]: The first node_num function is defined in `RoleMakerBase `, and the second node_num function is defined in `PaddleCloudRoleMaker `

**Pull request closed without merging**